### PR TITLE
Make config change detection more robust

### LINKED ISSUE #92: Config change detection gets a false positive when the watch expires
For various reasons, when you `Get()` an object, its `kind` and `apiVersion` are blank. When you receive the same object via a `Watch()`, those fields are present. Any time the watch expires (timeout, too old resource version), the config informer relists/rewatches and the update handler in `watchConfig` receives the current value of the config. Because of the mismatch described above, the code thinks there is a config change and shuts down.

### PR DESCRIPTION
**Pull request by ncdc**
Fix 2 issues with config change detection:
- Objects received via Get() don't have kind and apiVersion set, while
those from Watch() do, leading to false positives.
- Compare the unmodified config (prior to applying defaults) to the
updated one from Watch().

Fixes #92

Signed-off-by: Andy Goldstein <andy.goldstein@gmail.com>

### CODE REVIEW

File: pkg/cmd/server/server.go [hunk 1]
```diff
@@ -227,12 +236,25 @@ func (s *server) watchConfig(config *api.Config) {
...
 			if updated.Name != config.Name {
 				glog.V(5).Infof("config watch channel received other config %q", updated.Name)
 				return
 			}
 
+			// Objects retrieved via Get() don't have their Kind or APIVersion set. Objects retrieved via
+			// Watch(), including those from shared informer event handlers, DO have their Kind and
+			// APIVersion set. To prevent the DeepEqual() call below from considering Kind or APIVersion
+			// as the source of a change, set config.Kind and config.APIVersion to match the values from
+			// the updated Config.
```
Comment (jrnt30) [inline pkg/cmd/server/server.go]: Is this a known issue with CRDs in general or something specific to our implementation?  

And a more general question around the `APIVersion`, would that be used to signify a major upgrade in our configuration and should that be a consideration to simply ignoring it?

Comment (ncdc) [reply pkg/cmd/server/server.go]: > Is this a known issue with CRDs in general or something specific to our implementation?

This is actually the way that the Kubernetes API machinery works. It has nothing to do with CRDs specifically, as you'll see the same behavior with the client-go clientset and Namespaces, for example.

> And a more general question around the APIVersion, would that be used to signify a major upgrade in our configuration and should that be a consideration to simply ignoring it?

In general, a stable non-alpha non-beta API should always be backwards compatible as the software evolves. Breaking changes are only allowed in a newer version. This is the case with e.g. `storage.k8s.io/v1` - you can't remove fields or change the types of existing fields. Within an API group (`storage.k8s.io`), you can have multiple versions.

And here is where we get to something that is specific to CRDs: the Kubernetes apiserver only supports a single APIVersion per API group. I could go into more detail as to why, but I'll save that for another time, and simply say that we are "stuck" on `ark.heptio.com/v1`, and it's up to us to decide how and when we want to provide compatibility guarantees. For right now we're saying we're alpha, and we may change the API types, and you may have to wipe your existing data and start over. We'll try not to break things, but we may have to.

tl;dr: we'll only ever get Kind=Config and APIVersion=ark.heptio.com/v1, so it's safe to do this hack.

Comment (jrnt30) [reply pkg/cmd/server/server.go]: Thanks for taking the time with a detailed explanation, that's helpful context.

File: pkg/cmd/server/server.go [hunk 2]
```diff
@@ -125,13 +126,21 @@ func (s *server) run() error {
...
+	// separate object, and instead apply defaults to a clone.
+	copy, err := scheme.Scheme.DeepCopy(originalConfig)
+	if err != nil {
+		return err
+	}
+	config := copy.(*api.Config)
 	applyConfigDefaults(config)
 
-	s.watchConfig(config)
+	s.watchConfig(originalConfig)
```
Comment (davecheney) [inline pkg/cmd/server/server.go]: maybe move this above line 140 so we know at that point we are done with originalConfig 'cos its been handed off to watchConfig.

**Comment by skriss** [comment]
LGTM.
